cips/cip-x.md

```diff
@@ -0,0 +1,77 @@
+| cip | XX (assigned by Editors) |
+| - | - |
+| title | Header Pruning for Light Nodes |
+| description | Mechanism to retain a fixed range of headers instead of whole history |
+| author | Hlib Kanunnikov ([@Wondertan](https://github.com/Wondertan)) |
+| discussions-to | <https://forum.celestia.org/t/cip-header-pruning-for-lns/1958> |
+| status | Draft |
+| type | Standards Track |
+| category | Data Availability |
+| created | 2025-03-21 |
+
+## Abstract
+
+Currently, every data availability (DA) node type synchronizes all historical headers starting from genesis (or other statically configured
+historical header) until the subjectively initialized head of the chain. We change that by adding a way to sync a
+constant size range of headers instead of the whole history.
+
+## Motivation
+
+Light nodes (LNs) store the whole history of ExtendedHeaders in their entirety, including ValidatorSet, Commit and data availability header (DAH). The motivation
+is to decrease LN storage requirements by retaining only a configurable window of recent headers.
+
+## Specification
+
+In order to reference the last stored header, we introduce a notion of **Tail** header of the chain, complementing
+**Head** header. Tail header height is estimated with [HeaderPruningWindow](#parameters), block time and Head height.
+
+- On the first start of a node, the Tail is retrieved from trusted peers alongside with the Head, performing subjective
+initialization.
+  - Once retrieved, Tail becomes the point of trust for the initializing node where everything beforehand assumed valid
+ (unless syncing from genesis)
+  - Head, on the other hand, is not trusted and becomes a synchronization target for the node.
+  - All the headers from Tail to Head are then retrieved and verified using signature-based forward verification.
+- As the chain progresses, nodes continuously re-estimate Tail and cut off stored headers beyond the new Tail, retaining
+roughly [HeaderPruningWindow](#parameters) amount of headers.
+
+The estimation of Tail height is done as follows:
+
+```go
+func estimateTail(head, tail Header, blockTime, pruningWindow time.Duration) (height uint64) {
+	estimatedRange = (now + pruningWindow - tail.Time()) / blockTime
+	estimatedCutoffHeight = min(tail.Height() + estimatedRange, head.Height())
+	return estimatedCutoffHeight
+}
+```
+
+## Parameters
+
+| Parameter           | Proposed value | Description                                   | Changeable via Governance |
+|---------------------|----------------|-----------------------------------------------|---------------------------|
+| HeaderPruningWindow | 30 days        | Defines the length of time to retain headers  | No                        |
```

Comment:
this isnt really enforcable anywhere right? do we have a  way to distinguish between parameters that are enforceable as part of consensus and those that are just local configurations but are nonetheless important for security?

Reply:
Do you mean distinguishable in CIP process? I don't think we have it, but might be worth introducing.
cc @jcstein

Reply:
Yeh that's all I meant. Looks like parameters have moved out of CIP into https://celestiaorg.github.io/celestia-app/parameters.html so maybe there should just be a section there for the non consensus params

Reply:
> this isnt really enforcable anywhere right?

is it enforceable on celestia-node within the code @Wondertan ? are parameters like this best to leave documented in the code or should we have them publicly accessible elsewhere similar to the celestia-app specs?

> Looks like parameters have moved out of CIP into https://celestiaorg.github.io/celestia-app/parameters.html 

yeah, these have moved and CIP-13 is withdrawn.

Reply:
These types of parameter should be publicly specified as there are security and product implications, but there are not consensus/state machine enforcable.